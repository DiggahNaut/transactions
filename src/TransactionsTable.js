import React from 'react';
import {useTable, usePagination, useFilters, useGlobalFilter} from 'react-table';

// Компоненты из react-table оказались сложными для понимания с первого взгляда.
// Поэтому я решил сильно не углубляться в функциональность и остановился на "простом" использовании.

// Эта функция необходима для определения глобального фильтра. Её я скопировал из примера react-table.
function GlobalFilter({
  preGlobalFilteredRows,
  globalFilter,
  setGlobalFilter,
}) {
  return (
    <span>
      Поиск:{' '}
      <input
        className="table-filter"
        value={globalFilter || ''}
        onChange={e => {
          setGlobalFilter(e.target.value || undefined) // Set undefined to remove the filter entirely
        }}
        placeholder={`по всем записям`}
      />
    </span>
  )
}

// Данные берутся на основе data, которые хранятся в state у App.js.
// Невероятно, но факт. Передавая transactionsData из App.js и принимая data - данные перестают существовать.
// Передавая data и принимая data - всё работает исправно. Не знаю почему так.
function TransactionsTable({ data }) {
    // columns нужен для react-table.
    const columns = React.useMemo(
      () => [
        {
          Header: 'Список транзакций',
          columns: [
            {
              Header: 'Проект',
              accessor: 'transaction.project.name',
            },
            {
              Header: 'Способ оплаты',
              accessor: 'transaction.payment_method.name',
            },
            {
              Header: 'Пользователь',
              accessor: 'user.id',
            },
            {
              Header: 'Статус транзакции',
              accessor: 'transaction.status',
            }
          ],
        }
      ],
      []
    )

    // Используйте state и функции которые были получены от react-table для интерфейса таблицы.
    const {
      // Аргументы ниже нужны для react-table.
      getTableProps,
      getTableBodyProps,
      headerGroups,
      page,
      prepareRow,

      // Аргументы ниже нужны для пагинации react-table.
      canPreviousPage,
      canNextPage,
      pageOptions,
      pageCount,
      gotoPage,
      nextPage,
      previousPage,
      setPageSize,
      state: {pageIndex, pageSize}, // Не совсем понимаю разницу между этим state.

      // Аргументы ниже нужны для фильтра react-table.
      preGlobalFilteredRows,
      setGlobalFilter,
      flatColumns,
      state, // И этим state. В react-table не нашёл ответа. Так сделано в примере react-table.
    } = useTable(
      {
        columns,
        data,
        initialState: {pageIndex: 0}
      },
      useFilters,
      useGlobalFilter,
      usePagination
    )
  
    // Отрисовываем таблицу.
    return (
      <>
      <table className="table table-striped table-hover" {...getTableProps()}>
        <thead>
          {headerGroups.map(headerGroup => (
            <tr {...headerGroup.getHeaderGroupProps()}>
              {headerGroup.headers.map(column => (
                <th className="table-cell" {...column.getHeaderProps()}>{column.render('Header')}</th>
              ))}
            </tr>
          ))}
          <tr>
            <th colSpan={flatColumns.length} >
              <GlobalFilter
                preGlobalFilteredRows={preGlobalFilteredRows}
                globalFilter={state.globalFilter}
                setGlobalFilter={setGlobalFilter}
              />
            </th>
          </tr>
        </thead>

        <tbody {...getTableBodyProps()}>
          {page.map((row, i) => {
            prepareRow(row)
            return (
              <tr {...row.getRowProps()}>
                {row.cells.map(cell => {
                  return <td {...cell.getCellProps()}>{cell.render('Cell')}</td>
                })}
              </tr>
            )
          })}
        </tbody>
      </table>
      
      <div className="pagination">
        <div>
            <button type="button" className="btn btn-secondary" onClick={() => gotoPage(0)} disabled={!canPreviousPage}>
              {'<<'}
            </button>
              {' '}
            <button type="button" className="btn btn-secondary" onClick={() => previousPage()} disabled={!canPreviousPage}>
              {'<'}
            </button>
              {' '}
            <button type="button" className="btn btn-secondary" onClick={() => nextPage()} disabled={!canNextPage}>
              {'>'}
            </button>
              {' '}
            <button type="button" className="btn btn-secondary" onClick={() => gotoPage(pageCount - 1)} disabled={!canNextPage}>
              {'>>'}
            </button>
        </div>

        <div>
            Страница{' '}{pageIndex + 1} - {pageOptions.length}{' '}из{' '}{data.length}{' '}записей
        </div>

        <div>
          <select className="form-control"
            value={pageSize}
            onChange={e => {
              setPageSize(Number(e.target.value))
            }}
            >
            {[5, 10, 15, 20].map(pageSize => (
              <option key={pageSize} value={pageSize}>
                Показать {pageSize}
              </option>
            ))}
          </select>
        </div>
      </div>
    </>
    )
}

export default TransactionsTable