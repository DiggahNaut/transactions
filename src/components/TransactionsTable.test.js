import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import TransactionsTable from './TransactionsTable';
import transactionsData from '../transactions.json';

// В TransactionsTable в основном использован сторонний компонент. Хотелось бы узнать что можно протестировать тут и каким образом?
// И как вообще тестировать компоненты со сторонними библиотеками/компонентами?
// Мои предположения: Имитацию браузера с нажатием на кнопки, Проверка загрузки стороннего компонента,
// Проверка изменения количества строк в отображаемом списке при изменении количества из селектора,
// Проверка поиска и изменения таблицы с соответствующим поиском.
it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<TransactionsTable data={transactionsData}/>, div);
});

it("check TransactionsTable snapshot", () => {
    const TransactionsTableComponent = renderer.create(<TransactionsTable data={transactionsData}/>).toJSON();
    expect(TransactionsTableComponent).toMatchSnapshot();
});