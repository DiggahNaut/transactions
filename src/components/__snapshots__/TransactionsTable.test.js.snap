// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check TransactionsTable snapshot 1`] = `
Array [
  <table
    className="table table-striped table-hover"
  >
    <thead>
      <tr>
        <th
          className="table-cell"
          colSpan={4}
        >
          Список транзакций
        </th>
      </tr>
      <tr>
        <th
          className="table-cell"
          colSpan={1}
        >
          Проект
        </th>
        <th
          className="table-cell"
          colSpan={1}
        >
          Способ оплаты
        </th>
        <th
          className="table-cell"
          colSpan={1}
        >
          Пользователь
        </th>
        <th
          className="table-cell"
          colSpan={1}
        >
          Статус транзакции
        </th>
      </tr>
      <tr>
        <th
          colSpan={4}
        >
          <span>
            Поиск:
             
            <input
              className="table-filter"
              onChange={[Function]}
              placeholder="по всем записям"
              value=""
            />
          </span>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Red Sauce 3
        </td>
        <td>
          Credit/Debit Cards
        </td>
        <td>
          Jack Black
        </td>
        <td>
          created
        </td>
      </tr>
      <tr>
        <td>
          Red Sauce 3
        </td>
        <td>
          QIWI
        </td>
        <td>
          e.polyanina
        </td>
        <td>
          canceled
        </td>
      </tr>
      <tr>
        <td>
          Red Sauce 3
        </td>
        <td>
          Credit/Debit Cards
        </td>
        <td>
          Jack Black
        </td>
        <td>
          created
        </td>
      </tr>
      <tr>
        <td>
          Red Sauce 3
        </td>
        <td>
          Credit/Debit Cards
        </td>
        <td>
          lester.mosley
        </td>
        <td>
          canceled
        </td>
      </tr>
      <tr>
        <td>
          Red Sauce 3
        </td>
        <td>
          Credit/Debit Cards
        </td>
        <td>
          e.polyanina
        </td>
        <td>
          canceled
        </td>
      </tr>
      <tr>
        <td>
          Red Sauce 3
        </td>
        <td>
          Credit/Debit Cards
        </td>
        <td>
          lester.mosley
        </td>
        <td>
          canceled
        </td>
      </tr>
      <tr>
        <td>
          Red Sauce 3
        </td>
        <td>
          Credit/Debit Cards
        </td>
        <td>
          lester.mosley
        </td>
        <td>
          canceled
        </td>
      </tr>
      <tr>
        <td>
          Red Sauce 3
        </td>
        <td>
          Credit/Debit Cards
        </td>
        <td>
          lester.mosley
        </td>
        <td>
          canceled
        </td>
      </tr>
      <tr>
        <td>
          God of Peace
        </td>
        <td>
          Credit/Debit Cards
        </td>
        <td>
          john.uniqe
        </td>
        <td>
          canceled
        </td>
      </tr>
      <tr>
        <td>
          Super Farm
        </td>
        <td>
          MobileGo
        </td>
        <td>
          karina.newman
        </td>
        <td>
          created
        </td>
      </tr>
    </tbody>
  </table>,
  <div
    className="pagination"
  >
    <div>
      <button
        className="btn btn-secondary"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        &lt;&lt;
      </button>
       
      <button
        className="btn btn-secondary"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        &lt;
      </button>
       
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        &gt;
      </button>
       
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        &gt;&gt;
      </button>
    </div>
    <div>
      Страница
       
      1
       - 
      11
       
      из
       
      101
       
      записей
    </div>
    <div>
      <select
        className="form-control"
        onChange={[Function]}
        value={10}
      >
        <option
          value={5}
        >
          Показать 
          5
        </option>
        <option
          value={10}
        >
          Показать 
          10
        </option>
        <option
          value={15}
        >
          Показать 
          15
        </option>
        <option
          value={20}
        >
          Показать 
          20
        </option>
      </select>
    </div>
  </div>,
]
`;
